source/pip/src/interpreter.rs

```diff
@@ -821,64 +876,29 @@ fn first_unsupported_interop_ty(ty: &Ty) -> Option<&Ty> {
-    match ty {
-        Ty::Prim(prim_ty) => match prim_ty {
-            Prim::BigInt => Ok(Value::BigInt(obj.extract::<BigInt>(py)?)),
-            Prim::Bool => Ok(Value::Bool(obj.extract::<bool>(py)?)),
-            Prim::Double => Ok(Value::Double(obj.extract::<f64>(py)?)),
-            Prim::Int => Ok(Value::Int(obj.extract::<i64>(py)?)),
-            Prim::String => Ok(Value::String(obj.extract::<String>(py)?.into())),
-            Prim::Result => Ok(Value::Result(qsc::interpret::Result::Val(
-                obj.extract::<Result>(py)? == Result::One,
-            ))),
-            Prim::Pauli => Ok(Value::Pauli(match obj.extract::<Pauli>(py)? {
-                Pauli::I => fir::Pauli::I,
-                Pauli::X => fir::Pauli::X,
-                Pauli::Y => fir::Pauli::Y,
-                Pauli::Z => fir::Pauli::Z,
-            })),
-            Prim::Qubit | Prim::Range | Prim::RangeTo | Prim::RangeFrom | Prim::RangeFull => {
-                unimplemented!("primitive input type: {prim_ty:?}")
-            }
-        },
-        Ty::Tuple(tup) => {
-            if tup.len() == 1 {
-                let value = convert_obj_with_ty(py, obj, &tup[0]);
-                Ok(Value::Tuple(vec![value?].into()))
-            } else {
-                let obj = obj.extract::<Vec<PyObject>>(py)?;
-                if obj.len() != tup.len() {
-                    return Err(QSharpError::new_err(format!(
-                        "mismatched tuple arity: expected {}, got {}",
-                        tup.len(),
-                        obj.len()
-                    )));
-                }
-                let mut values = Vec::with_capacity(obj.len());
-                for (i, ty) in tup.iter().enumerate() {
-                    values.push(convert_obj_with_ty(py, &obj[i], ty)?);
+            .find(|t| first_unsupported_interop_ty(ctx, t).is_some()),
+        Ty::Array(ty) => first_unsupported_interop_ty(ctx, ty),
+        Ty::Udt(_, res) => {
+            let qsc::hir::Res::Item(item_id) = res else {
+                panic!("Udt should be an item");
+            };
+            let Some(udt) = ctx.udt_ty(item_id) else {
+                return Some(ty);
+            };
+
+            let Ok(fields) = collect_udt_fields(udt) else {
+                return Some(ty);
+            };
+
+            for field in fields {
+                if let Some(ty) = first_unsupported_interop_ty(ctx, field.1) {
+                    return Some(ty);
                 }
-                Ok(Value::Tuple(values.into()))
             }
+
+            None
         }
-        Ty::Array(ty) => {
-            let obj = obj.extract::<Vec<PyObject>>(py)?;
-            let mut values = Vec::with_capacity(obj.len());
-            for item in &obj {
-                values.push(convert_obj_with_ty(py, item, ty)?);
-            }
-            Ok(Value::Array(values.into()))
-        }
-        _ => unimplemented!("input type: {ty}"),
+        _ => Some(ty),
```

Comment:
Does it make sense to explicitly add the cases that result in unsupported tys here instead of `_`?